frontend/catalyst/from_plxpr.py

```diff
@@ -0,0 +1,374 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule defines a utility for converting plxpr into catalyst jaxpr.
+"""
+from dataclasses import dataclass, field
+from functools import partial
+from typing import Callable
+
+import jax
+from jax.extend.linear_util import wrap_init
+from pennylane.capture import AbstractOperator, AbstractMeasurement, qnode_prim
+
+
+from .device import extract_backend_info, get_device_capabilities
+from .jax_primitives import (
+    AbstractQreg,
+    compbasis_p,
+    expval_p,
+    func_p,
+    namedobs_p,
+    probs_p,
+    qalloc_p,
+    qdealloc_p,
+    qdevice_p,
+    qextract_p,
+    qinsert_p,
+    qinst_p,
+    sample_p,
+    state_p,
+    var_p,
+)
+from .utils.toml import ProgramFeatures
+
+measurement_map = {
+    "sample_wires": sample_p,
+    "expval_obs": expval_p,
+    "var_obs": var_p,
+    "probs_wires": probs_p,
+    "state_wires": state_p,
+}
+
+def _get_shapes_for(*measurements, shots=None, num_device_wires=0):
+    if jax.config.jax_enable_x64:
+        dtype_map = {
+            float: jax.numpy.float64,
+            int: jax.numpy.int64,
+            complex: jax.numpy.complex128,
+        }
+    else:
+        dtype_map = {
+            float: jax.numpy.float32,
+            int: jax.numpy.int32,
+            complex: jax.numpy.complex64,
+        }
+
+    shapes = []
+    if not shots:
+        shots = [None]
+
+    for s in shots:
+        for m in measurements:
+            shape, dtype = m.abstract_eval(shots=s, num_device_wires=num_device_wires)
+            shapes.append(jax.core.ShapedArray(shape, dtype_map.get(dtype, dtype)))
+    return shapes
+
+# pylint: disable=unidiomatic-typecheck
+def _read(var, env: dict):
+    return var.val if type(var) is jax.core.Literal else env[var]
+
+
+def _get_device_kwargs(device: "pennylane.devices.Device") -> dict:
+    """Calulcate the params for a device equation."""
+    features = ProgramFeatures(device.shots is not None)
+    capabilities = get_device_capabilities(device, features)
+    info = extract_backend_info(device, capabilities)
+    # Note that the value of rtd_kwargs is a string version of
+    # the info kwargs, not the info kwargs itself
```

Comment:
Yeah... I'm not sure why.

Reply:
Okay! @dime10 , do you have context here? I think it might be nice to comment on it :)

Reply:
Yes! This is just the easiest way to serialize the data since it is going through the IR. The alternative would be to provide typed attributes for each config option which is more work.

Reply:
Thanks for the info.  Adding that as a clarifying source code comment.